File: src/storage.rs
use std::path::PathBuf;
use tokio::fs;
use serde::{Deserialize, Serialize};
use chrono::{DateTime, Utc};
use crate::error::TrackerError;
use crate::models::StepInfo;

#[derive(Debug, Serialize, Deserialize)]
pub struct PipelineState {
    pub message_id: String,
    pub pr_number: u32,
    pub pr_title: String,
    pub author: String,
    pub repository: String,
    pub branch: String,
    pub steps: Vec<StepInfo>,
    pub pipeline_started_at: DateTime<Utc>,
}

#[derive(Debug)]
pub struct MessageStorage {
    file_path: PathBuf,
}

impl MessageStorage {
    pub fn new() -> Result<Self, TrackerError> {
        let current_dir = std::env::current_dir()?;
        let file_path = current_dir.join(".discord-pipeline-state");
        
        Ok(Self { file_path })
    }

    pub async fn save_pipeline_state(&self, state: &PipelineState) -> Result<(), TrackerError> {
        let json = serde_json::to_string_pretty(state)?;
        fs::write(&self.file_path, json).await?;
        Ok(())
    }

    pub async fn load_pipeline_state(&self) -> Result<Option<PipelineState>, TrackerError> {
        if !self.file_path.exists() {
            return Ok(None);
        }

        let content = fs::read_to_string(&self.file_path).await?;
        if content.trim().is_empty() {
            return Ok(None);
        }

        let state: PipelineState = serde_json::from_str(&content)?;
        Ok(Some(state))
    }

    pub async fn clear_pipeline_state(&self) -> Result<(), TrackerError> {
        if self.file_path.exists() {
            fs::remove_file(&self.file_path).await?;
        }
        Ok(())
    }

    // Legacy methods for backward compatibility
    pub async fn save_message_id(&self, message_id: &str) -> Result<(), TrackerError> {
        // Try to load existing state or create new one
        let mut state = self.load_pipeline_state().await?.unwrap_or_else(|| PipelineState {
            message_id: message_id.to_string(),
            pr_number: 0,
            pr_title: "Unknown".to_string(),
            author: "Unknown".to_string(),
            repository: "Unknown".to_string(),
            branch: "Unknown".to_string(),
            steps: Vec::new(),
            pipeline_started_at: Utc::now(),
        });
        
        state.message_id = message_id.to_string();
        self.save_pipeline_state(&state).await
    }

    pub async fn load_message_id(&self) -> Result<Option<String>, TrackerError> {
        if let Some(state) = self.load_pipeline_state().await? {
            Ok(Some(state.message_id))
        } else {
            Ok(None)
        }
    }

    pub async fn clear_message_id(&self) -> Result<(), TrackerError> {
        self.clear_pipeline_state().await
    }

    pub fn get_file_path(&self) -> &PathBuf {
        &self.file_path
    }
} 

File: src/storage.ts
import * as fs from 'fs/promises';
import * as path from 'path';
import { TrackerError } from './error';
import { StepInfo, PipelineState } from './models';
import { Storage } from './pipelineTracker';

// Internal pipeline state interface (matches PipelineTracker format)
interface InternalPipelineState {
	messageId: string;
	prNumber: number;
	prTitle: string;
	author: string;
	repository: string;
	branch: string;
	steps: StepInfo[];
	pipelineStartedAt: Date;
}

// Legacy format for backward compatibility
interface LegacyPipelineState {
	message_id: string;
	pr_number: number;
	pr_title: string;
	author: string;
	repository: string;
	branch: string;
	steps: StepInfo[];
	pipeline_started_at: Date;
}

/// File-based storage implementation that implements the Storage interface
export class FileStorage implements Storage {
  private filePath: string;

  constructor() {
    // Get current directory and construct file path
    // In Node.js, process.cwd() gives the current working directory
    const currentDir = process.cwd();
    this.filePath = path.join(currentDir, '.discord-pipeline-state');
  }

  /// Saves the current pipeline state to a file
  async savePipelineState(state: PipelineState): Promise<void> {
    try {
      const json = JSON.stringify(state, null, 2); // Pretty print JSON
      await fs.writeFile(this.filePath, json, 'utf-8');
    } catch (error: any) {
      throw TrackerError.fileSystemError(error);
    }
  }

  /// Loads the pipeline state from a file
  async loadPipelineState(): Promise<PipelineState | null> {
    try {
      const content = await fs.readFile(this.filePath, 'utf-8');
      // Check if content is empty AFTER reading it successfully
      if (content.trim() === '') {
        return null; // File is empty
      }
      const state: PipelineState = JSON.parse(content);
      return state;
    } catch (error: any) {
      // If file doesn't exist, return null
      if (error.code === 'ENOENT') {
        return null;
      }
      // If JSON parsing fails, it might be corrupted, treat as non-existent or throw
      throw TrackerError.jsonError(error);
    }
  }

  /// Clears the pipeline state file
  async clearPipelineState(): Promise<void> {
    try {
      await fs.unlink(this.filePath);
    } catch (error: any) {
      // If file doesn't exist, it's already cleared, so ignore error
      if (error.code !== 'ENOENT') {
        throw TrackerError.fileSystemError(error);
      }
    }
  }

  getFilePath(): string {
    return this.filePath;
  }
}

/// Legacy MessageStorage class for backward compatibility
export class MessageStorage {
  private filePath: string;

  constructor() {
    // Get current directory and construct file path
    // In Node.js, process.cwd() gives the current working directory
    const currentDir = process.cwd();
    this.filePath = path.join(currentDir, '.discord-pipeline-state');
  }

  /// Saves the current pipeline state to a file
  async savePipelineState(state: LegacyPipelineState): Promise<void> {
    try {
      const json = JSON.stringify(state, null, 2); // Pretty print JSON
      await fs.writeFile(this.filePath, json, 'utf-8');
    } catch (error: any) {
      throw TrackerError.fileSystemError(error);
    }
  }

  /// Loads the pipeline state from a file
  async loadPipelineState(): Promise<LegacyPipelineState | undefined> {
    try {
      const content = await fs.readFile(this.filePath, 'utf-8');
      // Check if content is empty AFTER reading it successfully
      if (content.trim() === '') {
        return undefined; // File is empty
      }
      const state: LegacyPipelineState = JSON.parse(content);
      return state;
    } catch (error: any) {
      // If file doesn't exist, return undefined
      if (error.code === 'ENOENT') {
        return undefined;
      }
      // If JSON parsing fails, it might be corrupted, treat as non-existent or throw
      throw TrackerError.jsonError(error);
    }
  }

  /// Clears the pipeline state file
  async clearPipelineState(): Promise<void> {
    try {
      await fs.unlink(this.filePath);
    } catch (error: any) {
      // If file doesn't exist, it's already cleared, so ignore error
      if (error.code !== 'ENOENT') {
        throw TrackerError.fileSystemError(error);
      }
    }
  }

  // Legacy methods for backward compatibility (simplified for TS)
  // These might need more robust handling if the old file format differs significantly

  async saveMessageId(messageId: string): Promise<void> {
    let state = await this.loadPipelineState();
    if (!state) {
      // Create a new state if none exists
      state = {
        message_id: messageId,
        pr_number: 0,
        pr_title: 'Unknown',
        author: 'Unknown',
        repository: 'Unknown',
        branch: 'Unknown',
        steps: [],
        pipeline_started_at: new Date(),
      };
    } else {
      state.message_id = messageId;
    }
    await this.savePipelineState(state);
  }

  async loadMessageId(): Promise<string | undefined> {
    const state = await this.loadPipelineState();
    return state?.message_id;
  }

  async clearMessageId(): Promise<void> {
    // This effectively clears the entire state, which is what clearPipelineState does
    await this.clearPipelineState();
  }

  getFilePath(): string {
    return this.filePath;
  }
}
